projects/noloan2/app/src/main/java/hasadna/noloan2/FirestoreClient.java

```diff
@@ -0,0 +1,88 @@
+package hasadna.noloan2;
+
+import android.support.annotation.NonNull;
+import android.util.Base64;
+
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.firestore.DocumentSnapshot;
+import com.google.firebase.firestore.FirebaseFirestore;
+import com.google.firebase.firestore.QuerySnapshot;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import hasadna.noloan2.protobuf.SMSProto.SmsMessage;
+
+
+public class FirestoreClient {
+  FirebaseFirestore client;
+  
+  String collection;
+  
+  public FirestoreClient(String collection) {
+    this.collection = collection;
+    client = FirebaseFirestore.getInstance();
+  }
+  
+  // Write the message to the Firestore
+  public void writeMassage(SmsMessage message) {
+    FirestoreElement element = encodeMessage(message);
+    client.collection(collection).add(element);
+  }
+  
+  QuerySnapshot snapshot;
+  
+  public ArrayList<SmsMessage> getMessages() {
+    ArrayList<SmsMessage> results = new ArrayList<>();
+    
+    client.collection(collection).get()
+      .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
+        @Override
+        public void onComplete(@NonNull Task<QuerySnapshot> task) {
+          if (task.isSuccessful()) {
+            snapshot = task.getResult();
+          } else {
+            //try again later
+          }
+        }
+      });
+    
+    // wait for the results
+    // TODO limit the time it take
+    while (snapshot == null) {
```

Comment:
This is a busy-wait and is considered bad because it keeps the CPU busy.
It's better to wait on a Task.

https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/Query.html#get()
The `get()` returns a `Task` object.
You can wait on it using `Tasks.await()`:
https://developers.google.com/android/reference/com/google/android/gms/tasks/Tasks.html#await(com.google.android.gms.tasks.Task%3CTResult%3E)

Reply:
I knew its busy waiting, I wanted to look for solution later.
I will look in the link you gave me and fix it